File: leetcode/leetcode/Solutions/Tree/CountGoodNodesinBinaryTree.swift
//
//  CountGoodNodesinBinaryTree.swift
//  leetcode
//
//  Created by Kai Wang on 8/29/22.
//  Copyright © 2022 Pirate. All rights reserved.
//

import Foundation

class CountGoodNodesinBinaryTree {
    func goodNodes(_ root: TreeNode?) -> Int {
        var count = 0

        func helper(node: TreeNode?, largestInPath: Int) {
            guard let node = node else {
                return
            }
            if node.val >= largestInPath {
                count += 1
            }
            helper(node: node.left, largestInPath: max(largestInPath, node.val))
            helper(node: node.right, largestInPath: max(largestInPath, node.val))
        }

        helper(node: root, largestInPath: Int.min)

        return count
    }

    func test() {
        let case1 = [3,1,4,3,nil,1,5]
        let case2 = [3,3,nil,4,2]
        let case3 = [1]
        let cases = [case1, case2, case3]
        for input in cases {
            let tree = TreeNode(values: input)
            print(goodNodes(tree))
        }
    }
}

File: leetcode/Python/Trees/CountGoodNodesinBinaryTree.py
from datastructure.binarytree import TreeNode


class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        def dfs(node: TreeNode, max_in_path: int) -> int:
            if not node:
                return 0
            res = 1 if node.val >= max_in_path else 0
            max_val = max(node.val, max_in_path)
            res += dfs(node.left, max_val)
            res += dfs(node.right, max_val)
            return res

        return dfs(root, root.val)
